### PR TITLE
Add gelu op

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Add gelu op

Differential Revision: D14814811

### CODE REVIEW

File: caffe2/python/operator_test/activation_ops_test.py [hunk 1]
```diff
@@ -215,3 +217,27 @@ def leaky_relu_ref(X):
...
+            ["X"],
+            ["Y"],
+            fast_gelu=fast_gelu,
+        )
+
+        def gelu_ref(X):
+            cdf = 0.5 * (1.0 + erf(X / np.sqrt(2.0)))
+            return (X * cdf,)
+
+        tol = 1e-3 if fast_gelu else 1e-4
```
Comment (houseroad) [inline caffe2/python/operator_test/activation_ops_test.py]: do we want to try 1e-5 tolerance in non fast case.

File: caffe2/operators/gelu_op.cc [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+namespace caffe2 {
+
+template <>
+template <typename T>
+bool GeluFunctor<CPUContext>::
+operator()(const int N, const T* X, T* Y, CPUContext* context) const {
+  ConstEigenVectorArrayMap<T> X_arr(X, N);
+  EigenVectorArrayMap<T> Y_arr(Y, N);
+  if (fast_gelu) {
```
Comment (houseroad) [inline caffe2/operators/gelu_op.cc]: add more comments? e.g., the equation of fast-gelu and gelu? maybe the link of the paper?

File: caffe2/operators/gelu_op.cc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  return cost;
+}
+
+} // namespace
+
+// Input: X, output: Y
+OPERATOR_SCHEMA(Gelu)
+    .NumInputs(1)
+    .NumOutputs(1)
+    .Arg("fast_gelu", "If true, use fast approximation")
```
Comment (houseroad) [inline caffe2/operators/gelu_op.cc]: maybe some rough perf and accuracy difference in two modes?

File: caffe2/operators/gelu_op.cc [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    const T* X,
+    T* dX,
+    CPUContext* context) const {
+  const int N = std::accumulate(
+      dY_dims.cbegin(), dY_dims.cend(), 1, std::multiplies<int>());
+  ConstEigenVectorArrayMap<T> dY_arr(dY, N);
+  ConstEigenVectorArrayMap<T> X_arr(X, N);
+  EigenVectorArrayMap<T> dX_arr(dX, N);
+  if (fast_gelu) {
+    constexpr T kAlpha = gelu_utils::kSqrt2 / gelu_utils::kSqrtPi;
```
Comment (houseroad) [inline caffe2/operators/gelu_op.cc]: equations here?

**Review by houseroad — COMMENTED** [review]
I think inplace case won't work with GELU. Shall we detect such case and error out if so?

I am also wondering what's the perf and accurary different between two modes (fast vs non-fast).

File: caffe2/python/operator_test/activation_ops_test.py (hunk 1 above)
Comment (BIT-silence) [reply caffe2/python/operator_test/activation_ops_test.py]: The default value is 1e-4
https://github.com/pytorch/pytorch/blob/master/caffe2/python/hypothesis_test_util.py#L563

Comment (houseroad) [reply caffe2/python/operator_test/activation_ops_test.py]: make sense. My concern is unnecessary. :-)

File: caffe2/operators/gelu_op.cc (hunk 2 above)
Comment (BIT-silence) [reply caffe2/operators/gelu_op.cc]: Done

**Review by houseroad — APPROVED** [review]

**Comment by houseroad** [comment]
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/40605/console

failing build is legit, please take a look

**Pull request closed without merging**